gms/gossiper.cc

```diff
@@ -125,10 +125,11 @@ class feature_enabler : public i_endpoint_state_change_subscriber {
     void on_restart(inet_address, endpoint_state) override {}
 };
 
-gossiper::gossiper(feature_service& features, db::config& cfg)
-        : _feature_service(features)
+gossiper::gossiper(feature_service& features, db::config& cfg, abort_source& as)
```

Comment:
I'm not sure interrupting all services at once is the right approach in general.

Shouldn't we rather have the stop operation (`gossiper::stop_gossiping()`) trigger the interruption of internal operations, on a private abort_source? So that we don't interrupt anything until it's really time to stop the component. Just theorizing now, there could be another component which is relying on the gossiper to complete its (the component's) operations, so we should first let that component to stop itself (which may want to wait for those operations to complete), and only then stop the gossiper.

Reply:
We already use a similar approach in `storage_proxy.cc` - take global `abort_source`, and use `sleep_abortable` with it everywhere.

I can see that all paths that pass through `sleep_abortable`s from gossiper.cc are called either from HTTP API (directly or indirectly), or from initialization. When initialization is aborted, it is desirable to stop waiting. As for HTTP API, it will just fail the request if user is unlucky enough to perform it during shutdown.

Reply:
Do you mean `storage_service`? I don't see abort_source usage in storage_proxy.cc. It's true that storage_service uses this approach, I still think it's a bad approach in general.

It's also true that this approach is not currently harmful given what operations it affects.

Reply:
Another argument for the alternative approach is that we want to stop the gossiper operations not only on ^C, but also when exception interrupts the boot process. Making `stop()` trigger abort handles both cases.

Reply:
Sorry, I meant `storage_service.cc`, my mistake.

I agree with you argument, I remade aborting so as to abort on a private `abort_source`. I left aborting on Ctrl+C `abort_source` in initialization code.